File: app/filters/color_filters.py
"""Color and Adjustment Filters."""

import cv2
import numpy as np
from typing import List

from app.filters.base_filter import ImageFilter
from app.pfm import PFMSetting, SettingType


class GrayscaleFilter(ImageFilter):
    @property
    def name(self) -> str: return "Grayscale"
    @property
    def category(self) -> str: return "Color"

    def process(self, image: np.ndarray) -> np.ndarray:
        if len(image.shape) == 3:
            return cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
        return image


class SaturationFilter(ImageFilter):
    @property
    def name(self) -> str: return "Saturation"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("saturation", "Saturation", SettingType.NUMBER, 1.0, 0.0, 3.0, 0.1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        if len(image.shape) != 3:
            return image
        sat = self.get("saturation")
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV).astype(np.float32)
        hsv[:, :, 1] *= sat
        hsv[:, :, 1] = np.clip(hsv[:, :, 1], 0, 255)
        return cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR)


class HueFilter(ImageFilter):
    @property
    def name(self) -> str: return "Hue Shift"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("hue", "Hue Shift", SettingType.INTEGER, 0, -180, 180, 1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        if len(image.shape) != 3:
            return image
        hue = self.get("hue")
        hsv = cv2.cvtColor(image, cv2.COLOR_BGR2HSV).astype(np.int16)
        hsv[:, :, 0] = (hsv[:, :, 0] + hue) % 180
        return cv2.cvtColor(hsv.astype(np.uint8), cv2.COLOR_HSV2BGR)


class GammaFilter(ImageFilter):
    @property
    def name(self) -> str: return "Gamma Correction"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("gamma", "Gamma", SettingType.NUMBER, 1.0, 0.1, 5.0, 0.1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        gamma = self.get("gamma")
        if gamma == 1.0: return image
        inv_gamma = 1.0 / gamma
        table = np.array([((i / 255.0) ** inv_gamma) * 255 for i in np.arange(0, 256)]).astype("uint8")
        return cv2.LUT(image, table)


class ExposureFilter(ImageFilter):
    @property
    def name(self) -> str: return "Exposure"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("exposure", "Exposure (EV)", SettingType.NUMBER, 0.0, -5.0, 5.0, 0.1)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        ev = self.get("exposure")
        if ev == 0.0: return image
        multiplier = 2 ** ev
        img = image.astype(np.float32) * multiplier
        return np.clip(img, 0, 255).astype(np.uint8)


class SepiaFilter(ImageFilter):
    @property
    def name(self) -> str: return "Sepia"
    @property
    def category(self) -> str: return "Color"

    def process(self, image: np.ndarray) -> np.ndarray:
        if len(image.shape) != 3:
            image = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
        kernel = np.array([[0.272, 0.534, 0.131],
                           [0.349, 0.686, 0.168],
                           [0.393, 0.769, 0.189]])
        sepia = cv2.transform(image, kernel)
        return np.clip(sepia, 0, 255).astype(np.uint8)


class CLAHEFilter(ImageFilter):
    @property
    def name(self) -> str: return "CLAHE"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("clip_limit", "Clip Limit", SettingType.NUMBER, 2.0, 0.1, 10.0, 0.1),
            PFMSetting("grid_size", "Grid Size", SettingType.INTEGER, 8, 2, 32, 2)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        clip = self.get("clip_limit")
        grid = self.get("grid_size")
        clahe = cv2.createCLAHE(clipLimit=clip, tileGridSize=(grid, grid))
        
        if len(image.shape) == 3:
            lab = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
            lab[:, :, 0] = clahe.apply(lab[:, :, 0])
            return cv2.cvtColor(lab, cv2.COLOR_LAB2BGR)
        else:
            return clahe.apply(image)


class TemperatureFilter(ImageFilter):
    @property
    def name(self) -> str: return "Temperature"
    @property
    def category(self) -> str: return "Color"

    def _define_settings(self) -> List[PFMSetting]:
        return [
            PFMSetting("temperature", "Temp", SettingType.NUMBER, 0.0, -100.0, 100.0, 1.0)
        ]

    def process(self, image: np.ndarray) -> np.ndarray:
        if len(image.shape) != 3: return image
        temp = self.get("temperature")
        if temp == 0.0: return image
        
        # Simple heuristic: add to red, subtract from blue
        img = image.astype(np.int16)
        img[:, :, 2] += int(temp) # R
        img[:, :, 0] -= int(temp) # B
        return np.clip(img, 0, 255).astype(np.uint8)

File: cpp/filters/color_filters.cpp
#include "filters/color_filters.h"
#include <opencv2/imgproc.hpp>
#include <cmath>

// ---------------------------------------------------------------------------
// GrayscaleFilter
// ---------------------------------------------------------------------------
cv::Mat GrayscaleFilter::process(const cv::Mat& image) {
    if (image.channels() == 3) {
        cv::Mat out;
        cv::cvtColor(image, out, cv::COLOR_BGR2GRAY);
        return out;
    } else if (image.channels() == 4) {
        cv::Mat out;
        cv::cvtColor(image, out, cv::COLOR_BGRA2GRAY);
        return out;
    }
    return image.clone();
}

// ---------------------------------------------------------------------------
// SaturationFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> SaturationFilter::defineSettings() const {
    return {
        { "saturation", "Saturation", SettingType::Number, 1.0, QVariant(), 0.0, 3.0, 0.0, 3.0, 0.1 }
    };
}

cv::Mat SaturationFilter::process(const cv::Mat& image) {
    if (image.channels() != 3) return image.clone();
    double sat = get("saturation").toDouble();
    if (std::abs(sat - 1.0) < 1e-4) return image.clone();

    cv::Mat hsv;
    cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);
    
    // Scale saturation (channel 1)
    std::vector<cv::Mat> channels;
    cv::split(hsv, channels);
    channels[1].convertTo(channels[1], CV_32F);
    channels[1] *= sat;
    channels[1].convertTo(channels[1], CV_8U);
    cv::merge(channels, hsv);

    cv::Mat out;
    cv::cvtColor(hsv, out, cv::COLOR_HSV2BGR);
    return out;
}

// ---------------------------------------------------------------------------
// HueFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> HueFilter::defineSettings() const {
    return {
        { "hue", "Hue Shift", SettingType::Integer, 0, QVariant(), -180, 180, -180, 180, 1 }
    };
}

cv::Mat HueFilter::process(const cv::Mat& image) {
    if (image.channels() != 3) return image.clone();
    int hue = get("hue").toInt();
    if (hue == 0) return image.clone();

    cv::Mat hsv;
    cv::cvtColor(image, hsv, cv::COLOR_BGR2HSV);
    
    // Shift Hue (channel 0)
    std::vector<cv::Mat> channels;
    cv::split(hsv, channels);
    
    channels[0].convertTo(channels[0], CV_16S);
    channels[0] += hue;
    
    // Wrap around 180
    for(int r = 0; r < channels[0].rows; r++) {
        short* row = channels[0].ptr<short>(r);
        for(int c = 0; c < channels[0].cols; c++) {
            if (row[c] < 0) row[c] += 180;
            if (row[c] >= 180) row[c] -= 180;
        }
    }
    
    channels[0].convertTo(channels[0], CV_8U);
    cv::merge(channels, hsv);

    cv::Mat out;
    cv::cvtColor(hsv, out, cv::COLOR_HSV2BGR);
    return out;
}

// ---------------------------------------------------------------------------
// GammaFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> GammaFilter::defineSettings() const {
    return {
        { "gamma", "Gamma", SettingType::Number, 1.0, QVariant(), 0.1, 5.0, 0.1, 5.0, 0.1 }
    };
}

cv::Mat GammaFilter::process(const cv::Mat& image) {
    double gamma = get("gamma").toDouble();
    if (std::abs(gamma - 1.0) < 1e-4) return image.clone();

    double invGamma = 1.0 / gamma;
    cv::Mat lut(1, 256, CV_8U);
    uchar* p = lut.ptr();
    for(int i = 0; i < 256; ++i) {
        p[i] = cv::saturate_cast<uchar>(std::pow(i / 255.0, invGamma) * 255.0);
    }
    
    cv::Mat out;
    cv::LUT(image, lut, out);
    return out;
}

// ---------------------------------------------------------------------------
// ExposureFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> ExposureFilter::defineSettings() const {
    return {
        { "exposure", "Exposure (EV)", SettingType::Number, 0.0, QVariant(), -5.0, 5.0, -5.0, 5.0, 0.1 }
    };
}

cv::Mat ExposureFilter::process(const cv::Mat& image) {
    double ev = get("exposure").toDouble();
    if (std::abs(ev) < 1e-4) return image.clone();

    double multiplier = std::pow(2.0, ev);
    cv::Mat imgFloat;
    image.convertTo(imgFloat, CV_32F);
    imgFloat *= multiplier;

    cv::Mat out;
    imgFloat.convertTo(out, CV_8U);
    return out;
}

// ---------------------------------------------------------------------------
// SepiaFilter
// ---------------------------------------------------------------------------
cv::Mat SepiaFilter::process(const cv::Mat& image) {
    cv::Mat img3;
    if (image.channels() == 1) {
        cv::cvtColor(image, img3, cv::COLOR_GRAY2BGR);
    } else {
        img3 = image.clone();
    }

    cv::Mat kernel = (cv::Mat_<float>(3, 3) <<
        0.272f, 0.534f, 0.131f,
        0.349f, 0.686f, 0.168f,
        0.393f, 0.769f, 0.189f);
    
    cv::Mat out;
    cv::transform(img3, out, kernel);
    out.convertTo(out, CV_8U);
    return out;
}

// ---------------------------------------------------------------------------
// CLAHEFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> CLAHEFilter::defineSettings() const {
    return {
        { "clip_limit", "Clip Limit", SettingType::Number, 2.0, QVariant(), 0.1, 10.0, 0.1, 10.0, 0.1 },
        { "grid_size", "Grid Size", SettingType::Integer, 8, QVariant(), 2, 32, 2, 32, 2 }
    };
}

cv::Mat CLAHEFilter::process(const cv::Mat& image) {
    double clip = get("clip_limit").toDouble();
    int grid = get("grid_size").toInt();

    auto clahe = cv::createCLAHE(clip, cv::Size(grid, grid));

    if (image.channels() == 3) {
        cv::Mat lab;
        cv::cvtColor(image, lab, cv::COLOR_BGR2Lab);
        
        std::vector<cv::Mat> channels;
        cv::split(lab, channels);
        clahe->apply(channels[0], channels[0]); // Apply to L channel
        cv::merge(channels, lab);
        
        cv::Mat out;
        cv::cvtColor(lab, out, cv::COLOR_Lab2BGR);
        return out;
    } else if (image.channels() == 1) {
        cv::Mat out;
        clahe->apply(image, out);
        return out;
    }
    return image.clone();
}

// ---------------------------------------------------------------------------
// TemperatureFilter
// ---------------------------------------------------------------------------
QVector<PFMSetting> TemperatureFilter::defineSettings() const {
    return {
        { "temperature", "Temp", SettingType::Number, 0.0, QVariant(), -100.0, 100.0, -100.0, 100.0, 1.0 }
    };
}

cv::Mat TemperatureFilter::process(const cv::Mat& image) {
    if (image.channels() != 3) return image.clone();
    double temp = get("temperature").toDouble();
    if (std::abs(temp) < 1e-4) return image.clone();

    cv::Mat img16;
    image.convertTo(img16, CV_16S);
    
    // Simple heuristic: OpenCV is BGR
    // B = 0, G = 1, R = 2
    // Positive temp adds red, removes blue
    std::vector<cv::Mat> channels;
    cv::split(img16, channels);
    channels[2] += temp; // Red
    channels[0] -= temp; // Blue
    cv::merge(channels, img16);

    cv::Mat out;
    img16.convertTo(out, CV_8U);
    return out;
}
